### DISCUSSION TITLE
How do I autouse a fixture only when another fixture has been requested?

### DISCUSSION THREAD

**Discussion by oakkitten** [pytest-dev/pytest#9851] [Q&A]
The title might sound like this is an XY problem, so let me explain.

I have a session scope fixture `app`, and a few other wide scope fixtures that depend on it. Also I have a fixture that does some app `cleanup` and I want it to run after each test. So I've got tests like this:

```python
def test_foo(app, cleanup):
    ...

def test_bar(app_with_profile_loaded, cleanup):
    ...
```

Now, I don't want to say `cleanup` in each test function arguments so I want to make it used automatically. However, if I do

```python
@pytest.fixture(autouse=True)
def cleanup():
    ...
```

then `cleanup` will also get called for tests that do not ask for `app`, which could break `cleanup`, and if I do

```python
@pytest.fixture(autouse=True)
def cleanup(app):
    ...
```

the app will be also constructed for tests that that do not ask for it, which is also problematic.

I could do the following, which should work,

```python
@pytest.fixture(autouse=True)
def cleanup(request):
    if "app" in request.fixturenames:
        ...
```

but I will then have `cleanup` for every test in `--setup-plan`. Is there a better way?

**Comment by brizjin** [comment]
Probably, your app fixture is session level. That is why you need separate fixture, function level, which will clean everything after each test and do no create app each test. You can solve it, with creating other function level fixture, which will sent app session level and then clean app everything. Then use app-function level everywhere in test it needs, and it will cleanup after test and do not destroy app object. And you will not have autouse fixtures, tests which no need app will not use it:
```
@fixture(scope='session')
def app_session():
    app = create_app()
    with app.app_context():
        yield app


@fixture
def app(app_session):
    yield app_session
    cleanup()
```

**Reply by oakkitten** [reply]
Hum, that works, but then I would have to define similar constructs for all fixtures that depend on `app`. I think *I* would prefer the `request` trick to this.
